src/sql/parts/notebook/models/notebookModel.ts

```diff
@@ -711,45 +790,41 @@ export class NotebookModel extends Disposable implements INotebookModel {
 	}
 
 	/**
-	 * Set _providerId and _activeClientSession based on a kernelSpec representing new kernel
-	 * @param kernelSpec KernelSpec for new kernel
+	 * Set _providerId and start session if it is new provider
+	 * @param displayName Kernel dispay name
 	 */
-	private setProviderIdForKernel(kernelSpec: nb.IKernelSpec): void {
-		if (!kernelSpec) {
-			// Just use the 1st non-default provider, we don't have a better heuristic
-			let notebookManagers = this._notebookOptions.notebookManagers.filter(manager => manager.providerId !== DEFAULT_NOTEBOOK_PROVIDER);
-			if (!notebookManagers.length) {
-				notebookManagers = this._notebookOptions.notebookManagers;
-			}
-			if (notebookManagers.length > 0) {
-				this._providerId = notebookManagers[0].providerId;
-			}
-		} else {
-			let sessionManagerFound: boolean = false;
-			for (let i = 0; i < this.notebookManagers.length; i++) {
-				if (this.notebookManagers[i].sessionManager && this.notebookManagers[i].sessionManager.specs && this.notebookManagers[i].sessionManager.specs.kernels) {
-					let index = this.notebookManagers[i].sessionManager.specs.kernels.findIndex(kernel => kernel.name === kernelSpec.name);
-					if (index >= 0 && this._clientSessions && this._clientSessions.length > 0) {
-						if (this._activeClientSession) {
-							this._oldClientSession = this._activeClientSession;
-						}
-						this._activeClientSession = this._clientSessions[i];
-						if (this.notebookManagers[i].providerId !== this._providerId) {
-							this._providerId = this.notebookManagers[i].providerId;
-							this._onProviderIdChanged.fire(this._providerId);
+	private async setProviderIdAndStartSession(displayName: string): Promise<void> {
+		if (displayName) {
+			if (this._activeClientSession && this._activeClientSession.isReady) {
+				this._oldKernel = this._activeClientSession.kernel;
+				let standardKernel = this.getStandardKernelFromDisplayName(displayName);
+				if (standardKernel && this._oldKernel && this._oldKernel.name !== standardKernel.name) {
+					if (this._kernelDisplayNameToNotebookProviderIds.has(displayName)) {
+						let providerId = this._kernelDisplayNameToNotebookProviderIds.get(displayName);
+						if (providerId) {
+							if (providerId !== this._providerId) {
+								this._providerId = providerId;
+								this._onProviderIdChanged.fire(this._providerId);
+
+								await this.shutdownActiveSession();
+
+								try {
+									let manager = this.getNotebookManager(providerId);
+									if (manager) {
+										await this.startSession(manager, displayName);
+									} else {
+										throw new Error(localize('ProviderNoManager', 'Can\'t find notebook manager for provider {0}', providerId));
```

Comment:
Apparently we're meant to use `"` instead of `'` for the actual localized message. remember to change `Can\'t` to `Can't` once you do this.

Reply:
fixed.